accepted/future-releases/0546-patterns/feature-specification.md

```diff
@@ -1884,27 +1884,51 @@ var [int a, b] = <num>[1, 2];   // List<num>.
 
 To type check a pattern `p` being matched against a value of type `M`:
 
-*   **Logical-or** and **logical-and**: Type check each branch using `M` as the
-    matched value type.
+*   **Logical-or**: Type check each branch using `M` as the
```

Comment:
Here is an example which shows that the promotion does not take place, at least in some cases: https://github.com/dart-lang/language/issues/2912.

Reply:
Just learned that this kind of analysis _does_ in fact occur. Changed the text to require that the associated promotions are taken into account.

Reply:
Commented on the issue. It seems we *do* do promotion in `||` patterns.

Which is great, I would generally expect/prefer if `patternTest1 || patternTest2` works the same way on the matched value "variable" as `test1 || test2` works on a similar test on a real variable. That is `== null || > 0` should work the same on a matched value of type `int` as `x == null || x > 0` will on a variable of type `int`.

What that implies about promotion *after* the full case pattern test is secondary (but preferably predictable in some way), but *internally* in a single composite pattern test we should not throw away information deduced by the first part of the test, whether that information comes from the test failing or succeeding. 

Not taking information from one pattern to another applies to *separate* cased of a switch. We don't want:
```dart
switch (maybeInt) {
   case (> 0)!: ....
   case var y: ... 
}
```
to take the `!` promotion from the first case pattern into the second case pattern.
That would only work because the `!` dominates all exits of the first pattern, but if it had been was `otherPattern && (> 0)!`, there might be an exit path which doesn't do `!`.

Or in other words, we don't treat a failing case pattern to have control flows going out of the pattern, from somewhere inside, and into then next case.
It either succeeds entirely, or it fails providing no information to the further switches, other than the checks it may have exhausted. 
(Exhausting union types _can_ promote the original variable or value:
```dart
    int? q = 2 as dynamic;
    switch (q) {
      case == null: q..log;
      case _ when q.log == -1: {}  // Prints "2:int @ int", q promoted to `int` here.
      case var x: x..log; q..log;  // Prints "2:int @ int"  twice
    }
```
Wonderful!)

Because of that, we can't propagate promotions happening inside the case to later cases. There is still control low *inside* the individual case patterns.